Book authorization

TatianaPan opened the pull request:
With book authorization logged in user can create, edit and destroy books in his own library only.

I have also added button `Library` , so now logged in user can see other users' list of books.

Branch is deployed to staging https://book-swap-app-staging.herokuapp.com/

ahorincar commented on test/policies/book_policy_test.rb:
```diff
@@ -0,0 +1,31 @@
+
+    sign_in user
+
+    refute_permit(user, book, :new?)
```
I think we can remove `refute_permit(user, book, :new?)` from here, since anyone can access the book form.

TatianaPan replied:
same as previous comment

ahorincar commented on app/views/books/index.html.haml:
```diff
@@ -1,4 +1,7 @@
+- if @user == current_user
+  %h1 My library
+- else
+  %h1= "#{@user.decorate.display_full_name}'s  library"
```
I logged in on staging via my Google account, and for some reason, when I go to someone else's library, their full name is blank. Do you know why this happens?

TatianaPan replied:
so the problem is that they didn't provide their names..  I think it could be handled in different ways: 
- after sign  up I should redirect user to `My profile` page with prompt 'Please fill in your name'
- render user's email instead, but in the future I wouldn't like to make users' emails public

TatianaPan replied:
as we have discussed, I'll change registration form so that all users will have to provide their names.

TatianaPan commented on app/policies/application_policy.rb:
```diff
@@ -0,0 +1,49 @@
+class ApplicationPolicy
```
as we have discussed, I deleted this file and included `def initialize` in `book_policy.rb`

TatianaPan commented on app/policies/book_policy.rb:
```diff
@@ -0,0 +1,21 @@
+class BookPolicy < ApplicationPolicy
+  def new?
+    record.user == user
```
as we've discussed, it's better to leave this policy, so that other users cannot access other's user  new form changing URL

TatianaPan commented:
> Looks good, I've added a few minor comments/suggestions.
> 
> I also think it'd be good to add some controller tests for the books controller, which check that an unauthorized user cannot `post`, `put` and `delete` on behalf of another user (aka they get a 302 error when firing the request).

Thank you very much for review @ahorincar ! I have added tests for controller.

 I have also had to change `create`method in `BooksController`, because before I used `book.create(books_params)`and thus new book has been saved _before_ authorization.  No I use `book.new(book_params)`, then `authorize @book`, then `book.save`.

ahorincar commented on test/controllers/books_controller_test.rb:
```diff
@@ -55,6 +73,17 @@ class BooksControllerTest < ActionDispatch::IntegrationTest
+
+    sign_in user
+
+    patch user_book_url(book.user, book), params: { book: { status: 'reserved' } }
```
Not a big deal, but you can also use the `assert_no_changes` helper here. E.g.
```ruby
assert_no_changes, 'book.status' do
  patch ...
end
```

TatianaPan replied:
Good idea! I changed it.

TatianaPan commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Add authorisation for Book](https://trello.com/c/MqYVXCSf/27-add-authorisation-for-book)
